python/xir/program.py

```diff
@@ -567,6 +567,65 @@ def serialize(self, minimize: bool = False) -> str:
             return strip(res_script)
         return res_script
 
+    @staticmethod
+    def merge(*programs) -> "XIRProgram":
+        """Merges one or more XIR programs into a new XIR program.
+
+        The merged XIR program is formed by concatenating the given XIR programs
+        in the order they are passed to the function. Warnings may be issued for
+        duplicate declarations, gates, includes, operators, and options.
+
+        Args:
+            programs (XIRProgram): XIR programs to merge
+
+        Returns:
+            XIRProgram: the merged XIR program
+
+        Raises:
+            ValueError: if no XIR programs are provided or if at least two XIR
+                programs have different versions or float settings
+        """
+        if len(programs) == 0:
+            raise ValueError("Merging requires at least one XIR program.")
+
+        version = programs[0].version
+        if any(program.version != version for program in programs):
+            raise ValueError("XIR programs with different versions cannot be merged.")
+
+        use_floats = programs[0].use_floats
+        if any(program.use_floats != use_floats for program in programs):
+            raise ValueError("XIR programs with different float settings cannot be merged.")
+
+        result = XIRProgram(version=version, use_floats=use_floats)
+
+        for program in programs:
+            for called_function in program.called_functions:
+                result.add_called_function(called_function)
+
+            for key, decls in program.declarations.items():
+                for decl in decls:
+                    result.add_declaration(key, decl)
+
+            for name, gate in program.gates.items():
+                result.add_gate(name, **gate)
+
+            for include in program.includes:
+                result.add_include(include)
+
+            for name, operator in program.operators.items():
+                result.add_operator(name, **operator)
+
+            for name, value in program.options.items():
+                result.add_option(name, value)
+
+            for statement in program.statements:
+                result.add_statement(statement)
+
+            for variable in program.variables:
+                result.add_variable(variable)
+
```

Comment:
Just looking through the warnings that are given if, e.g., having a gate defined more than once.
What is the motivation for raising a warning rather than an error?

Reply:
Excellent question! The motivation here is twofold (@thisac can correct me if I'm wrong):
1. Improve the flexibility of XIR programs (i.e., be liberal in what the parser accepts).
2. Make it easier to compose XIR programs (especially if they they are authored by different vendors).

My personal opinion is that borderline ambiguous syntax is fine so long as the parser is deterministic the user is aware of the potential confusion in their code.